alan-turing-institute/scivision
# Add ability to load data plugins

Changes on this scivision branch enable us to add "data plugins" which is the name I'm giving for Python packages that are set up specifically as middleware code between scivision and a data resource that can be loaded via an existing Python API.

I have set up an example data plugin, which is a python package that loads Sentinel-2 Cloud-Optimized GeoTIFF images via the `odc-stac` package: https://github.com/alan-turing-institute/scivision_sentinel2_stac 

### To test

Have a play with the `scivision_sentinel2_stac` data plugin:

1. check out this branch in the scivision repo and run `pip install -e .` to get the changes
2. look at the README on https://github.com/alan-turing-institute/scivision_sentinel2_stac and follow the instructions to open the notebook
3. You can see the args that are available [here](https://github.com/alan-turing-institute/scivision_sentinel2_stac/blob/main/scivision_sentinel2_stac/data.py) -  Let me know your thoughts and you could either comment here or open some issues in that repo.

I'm thinking I should document the API for this plugin in the plugin repo (not in scivision), i.e. what the arguments for the `get_images` func are and how to use via scivision `load_dataset` in the repo plugin itself e.g. in the README and/or via the example notebook. Perhaps the convention for data plugins should be to have their own documentation as a pre-requisite for inclusion in scivision. Otherwise it's not clear how a user would know what the arguments are when they load the data plugin (in the notebook where I do `data.load_data`).

### Reviewers

One thing that's not super nice about this that `load_dataset` can now return different object types depending on the input. The advantage of this is that we can be flexible with the returned data type of any data plugins, which will be useful as different datasources will inevitably come in different formats, and we can still use a single function. What do you think about this?

### TODO

- [x] Add `scivision_sentinel2_stac` as a datasource in the scivision catalog
    - [x] make sure catalog entry is accurate
- [x] Add test that shows data can be loaded via data plugin
- [ ] Document the API for this plugin in the plugin repo (not in scivision)
- [ ] Update docs to explain how to add and use a data plugin
     - [ ] Add a section about how the data plugin config differs [here](https://scivision.readthedocs.io/en/latest/data_repository_template.html#data-config-file)
     - [ ] Update the [API](https://scivision.readthedocs.io/en/latest/api.html) page
